File: backend/app/services/rate_limit.py
"""Rate limiting & the global daily job cap (plan §5) — atomic Postgres
counters via INSERT ... ON CONFLICT ... DO UPDATE ... RETURNING, so the
check-and-increment is race-free without a separate read-then-write step.

Deliberately explicit per-endpoint checks (not blanket middleware) — see
plan §5's rationale: cheap read endpoints shouldn't be throttled, and the
costly endpoints stay easy to audit.
"""

from datetime import UTC, datetime

from fastapi import HTTPException, status
from sqlalchemy.dialects.postgresql import insert
from sqlalchemy.orm import Session

from ..models import GlobalJobCounter, RateLimitCounter


def check_and_increment_ip(db: Session, ip: str, limit_per_hour: int) -> None:
    window_start = _truncate_to_hour(datetime.now(UTC))
    _check_and_increment(db, scope=f"ip:{ip}", window_start=window_start, limit=limit_per_hour)


def check_and_increment_user(db: Session, user_id: str, limit_per_day: int) -> None:
    window_start = _truncate_to_day(datetime.now(UTC))
    _check_and_increment(db, scope=f"user:{user_id}", window_start=window_start, limit=limit_per_day)


def check_and_increment_global_daily(db: Session, max_jobs_per_day: int) -> None:
    """The central backstop on total GPU spend (plan §5) — independent of
    user/IP identity, checked only when a job is actually about to launch.
    """
    day = _truncate_to_day(datetime.now(UTC))

    stmt = (
        insert(GlobalJobCounter)
        .values(day=day, jobs_started=1)
        .on_conflict_do_update(
            index_elements=[GlobalJobCounter.day],
            set_={"jobs_started": GlobalJobCounter.jobs_started + 1},
        )
        .returning(GlobalJobCounter.jobs_started)
    )
    new_count = db.execute(stmt).scalar_one()
    db.commit()

    if new_count > max_jobs_per_day:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail="Daily processing limit reached — try again tomorrow.",
        )


def _check_and_increment(db: Session, *, scope: str, window_start: datetime, limit: int) -> None:
    stmt = (
        insert(RateLimitCounter)
        .values(scope=scope, window_start=window_start, count=1)
        .on_conflict_do_update(
            index_elements=[RateLimitCounter.scope, RateLimitCounter.window_start],
            set_={"count": RateLimitCounter.count + 1},
        )
        .returning(RateLimitCounter.count)
    )
    new_count = db.execute(stmt).scalar_one()
    db.commit()

    if new_count > limit:
        raise HTTPException(
            status_code=status.HTTP_429_TOO_MANY_REQUESTS,
            detail="Rate limit exceeded — please slow down.",
        )


def _truncate_to_hour(dt: datetime) -> datetime:
    return dt.replace(minute=0, second=0, microsecond=0)


def _truncate_to_day(dt: datetime) -> datetime:
    return dt.replace(hour=0, minute=0, second=0, microsecond=0)

File: web/lib/server/rateLimit.ts
import { sql } from "drizzle-orm";

import { getDb } from "./db";
import { globalJobCounters, rateLimitCounters } from "./db/schema";
import { HttpError } from "./httpError";

/**
 * Rate limiting & the global daily job cap.
 *
 * Counters are incremented with a single
 * `INSERT ... ON CONFLICT ... DO UPDATE SET count = count + 1 RETURNING count`,
 * so the check-and-increment is race-free without a read-then-write step. The
 * `set` clause must keep referencing the column, never a JavaScript value —
 * AGENTS.md has the race that reopens, and how to check the SQL Postgres
 * actually received.
 *
 * Checks are per-endpoint rather than blanket middleware: cheap reads
 * shouldn't be throttled, and the costly endpoints stay easy to audit.
 */

export async function checkAndIncrementIp(ip: string, limitPerHour: number): Promise<void> {
  await checkAndIncrement(`ip:${ip}`, truncateToHour(new Date()), limitPerHour);
}

export async function checkAndIncrementUser(userId: string, limitPerDay: number): Promise<void> {
  await checkAndIncrement(`user:${userId}`, truncateToDay(new Date()), limitPerDay);
}

/**
 * The central backstop on total GPU spend — independent of
 * user/IP identity, checked only when a job is actually about to launch.
 */
export async function checkAndIncrementGlobalDaily(maxJobsPerDay: number): Promise<void> {
  const day = truncateToDay(new Date());

  const [counter] = await getDb()
    .insert(globalJobCounters)
    .values({ day, jobsStarted: 1 })
    .onConflictDoUpdate({
      target: globalJobCounters.day,
      set: { jobsStarted: sql`${globalJobCounters.jobsStarted} + 1` },
    })
    .returning({ jobsStarted: globalJobCounters.jobsStarted });

  if (counter.jobsStarted > maxJobsPerDay) {
    throw new HttpError(503, "Daily processing limit reached — try again tomorrow.");
  }
}

async function checkAndIncrement(scope: string, windowStart: Date, limit: number): Promise<void> {
  const [counter] = await getDb()
    .insert(rateLimitCounters)
    .values({ scope, windowStart, count: 1 })
    .onConflictDoUpdate({
      target: [rateLimitCounters.scope, rateLimitCounters.windowStart],
      set: { count: sql`${rateLimitCounters.count} + 1` },
    })
    .returning({ count: rateLimitCounters.count });

  if (counter.count > limit) {
    throw new HttpError(429, "Rate limit exceeded — please slow down.");
  }
}

function truncateToHour(dt: Date): Date {
  const out = new Date(dt);
  out.setUTCMinutes(0, 0, 0);
  return out;
}

function truncateToDay(dt: Date): Date {
  const out = new Date(dt);
  out.setUTCHours(0, 0, 0, 0);
  return out;
}
